### PR TITLE
Added groundwork for events

### PR DESCRIPTION
**Pull request by dktrotti**

### CODE REVIEW

File: Shanghai Blood Sports/Assets/Scripts/Events/EventHandler.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Assets.Scripts.Events {
+    public class EventHandler {
+        public void onEvent<EVENT, ARGS>(EVENT e)
+            where EVENT : Event<ARGS>
+            where ARGS : EventArgs {
+            if (e is HitEvent hit) {
```
Comment (Loonman) [inline Shanghai Blood Sports/Assets/Scripts/Events/EventHandler.cs]: https://stackoverflow.com/questions/298976/is-there-a-better-alternative-than-this-to-switch-on-type/299001#299001

Still not sold on if/switch case but this is probably more elegant

File: Shanghai Blood Sports/Assets/Scripts/Events/Event.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        /// <summary>
+        /// Executes the appropriate handlers.
+        /// </summary>
+        /// <returns>The modified arguments, or null if the event was cancelled.</returns>
+        public ARGS Execute() {
+            if (cancelled) {
+                cancelHandlers.ForEach(f => f(Args));
+                return null;
+            } else {
+                return resolveHandlers.Aggregate(Args, (args, f) => f(args));
```
Comment (Loonman) [inline Shanghai Blood Sports/Assets/Scripts/Events/Event.cs]: Does this guarantee order of evaluation? Do we care about order of evaluation?

File: Shanghai Blood Sports/Assets/Scripts/Events/BlockEvent.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Assets.Scripts.Events {
+    public class BlockEventArgs : EventArgs { }
+
+    public class BlockEvent : Event<BlockEventArgs> {
```
Comment (Loonman) [inline Shanghai Blood Sports/Assets/Scripts/Events/BlockEvent.cs]: Braces 😐

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/inside-a-program/coding-conventions

File: Shanghai Blood Sports/Assets/Scripts/Character.cs [hunk 4]
```diff
@@ -17,12 +17,14 @@ public class Character : MonoBehaviour
     }
 
     // Start is called before the first frame update
-    void Start() {
+    void Start()
+    {
```
Comment (Loonman) [inline Shanghai Blood Sports/Assets/Scripts/Character.cs]: https://www.youtube.com/watch?v=ahd4UOLlSaI

File: Shanghai Blood Sports/Assets/Scripts/Events/Event.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Assets.Scripts.Events {
+    public class EventArgs { }
+
+    public class Event<ARGS> where ARGS : EventArgs {
+        /// <summary>
```
[the remark this answers is not in the document]
Comment (dktrotti) [reply Shanghai Blood Sports/Assets/Scripts/Events/Event.cs]: That's what visual studio does if you type `///`. I don't think we need much documentation, so this should be fine.

File: Shanghai Blood Sports/Assets/Scripts/Events/Event.cs (hunk 2 above)
Comment (dktrotti) [reply Shanghai Blood Sports/Assets/Scripts/Events/Event.cs]: Hopefully we don't care about the order of operation, although we could apply some sorting if we're seeing non-deterministic behaviour.

**Pull request merged**